### ISSUE TITLE
getopt: passing more than one wrong arguments causes return with incremented `optwhere`

### ISSUE THREAD

**Issue by damianloew** [phoenix-rtos/phoenix-rtos-project#521]
- Here is the case, where it's problematic:
 
![Screenshot from 2022-10-06 17-08-32](https://user-images.githubusercontent.com/77304431/194349698-997e607f-b914-4db8-bc63-c90ca0149760.png)

 - During the second history call `optwhere` is set to 2, so instead of checking `c` termination character (0) is checked

**Comment by damianloew** [comment]
> 

The `optind` variable is set here and I verified that it's called during every history command call
https://github.com/phoenix-rtos/phoenix-rtos-utils/blob/6d6db73e493462d4720ba268ff1ce7150ab7978e/psh/pshapp/pshapp.c#L1112

**Comment by damianloew** [comment]
> * Here is the case, where it's problematic:
> 
> ![Screenshot from 2022-10-06 17-08-32](https://user-images.githubusercontent.com/77304431/194349698-997e607f-b914-4db8-bc63-c90ca0149760.png)
> 
> * During the second history call `optwhere` is set to 2, so instead of checking `c` termination character (0) is checked

Providing I understand correctly:

- We can reproduce the whole issue by the following simple program:
```C
int main(int argc, char *argv[])
{
    char c;

    if (argc >= 2) {
        /* First history call */
        while ((c = getopt(argc, argv, "ch")) != -1) {
            if (c =='c') {
                printf("c\n");
            }
            else {
                printf("default, c = %d -> break\n", c);
                break;
            }
        }
        /* ------------------ */

        optind = 1;
        printf("Second call\n");

        /* Second history call */
        while ((c = getopt(argc, argv, "ch")) != -1) {
            if (c =='c') {
                printf("c\n");
            }
            else {
                printf("default, c = %d -> break\n", c);
                break;
            }
        }
        /* ------------------ */
    }

    return 0;
}
```

 - Result both for host and Phoenix are similar:
 
![Screenshot from 2022-10-06 20-05-17](https://user-images.githubusercontent.com/77304431/194386641-32ec2845-4614-4df6-8430-834c6eb05b66.png)

![Screenshot from 2022-10-06 20-13-40](https://user-images.githubusercontent.com/77304431/194388066-7068ed39-9453-41d4-8902-c66afc9e2a83.png)

I suppose that we can't change the value of `optind`, because in fact `optind` is stored in static structure `_getopt_data` and it's redefined to 2. But I think that if we could the behavior would be the same - on host we have sth similar to `optwhere` - `__nextchar` in `_getoptdata` structure: https://codebrowser.dev/glibc/glibc/posix/getopt.c.html#getopt_data

 - To sum up: `getopt()` assumes that one program will use getopt one time and static global variables will be reset before that. It works on host, because for example history is separate program, but does not work on our system, where all applets are called in one `psh` program (static global variables, like `optwhere` are not restored) (?)

**Comment by kemonats** [comment]
I hadn't noticed.

Probable solution:
in `getopt()` you need to add:
```
	if (optind <= 0) {
		optind = 1;
		optwhere = 1;
	}

	if (argc == 0 || argv == NULL || optind >= argc)
		return -1;

```
and to reset `getopt()` use `0` instead `1`.
```
optind = 0;
```
Change in the test program from `optind = 1` on `optind = 0` and check on the host.

**Comment by damianloew** [comment]
> I hadn't noticed.
> 
> Probable solution: in `getopt()` you need to add:
> 
> ```
> 	if (optind <= 0) {
> 		optind = 1;
> 		optwhere = 1;
> 	}
> 
> 	if (argc == 0 || argv == NULL || optind >= argc)
> 		return -1;
> ```
> 
> and to reset `getopt()` use `0` instead `1`.
> 
> ```
> optind = 0;
> ```
> 
> Change in the test program from `optind = 1` on `optind = 0` and check on the host.

I suppose it won't help, because it's the matter of `optwhere`, not `optind`. I see that You tested it, what were the results?

**Comment by kemonats** [comment]
If we add the condition proposed above to the `getopt()` function and use `optind = 0` ( rather than `optind = 1`) to reset getopt(), it will work properly (`optind` and `optwhere` will be set to `1` and `getopt()` will work as if it were run for the first time). 
In addition, the condition will protect against an out-of-range array index.

**Comment by damianloew** [comment]
> 

Yes, You are right, it's possible by doing it this way. I was wrong about it, thanks for the PR :)

**Comment by nalajcie** [comment]
Fixed by https://github.com/phoenix-rtos/libphoenix/pull/208

**Comment by damianloew** [comment]
The case from issue hasn't been resolved yet. The following change is also needed:

https://github.com/phoenix-rtos/phoenix-rtos-utils/pull/157

I would rather close this issue after merging the PR above

**Comment by nalajcie** [comment]
no, this issue is about `getopt` not `psh` - which is now fixed
